examples/gadgets/simple_trace_open/main.go

```diff
@@ -0,0 +1,93 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/datasource"
+	igjson "github.com/inspektor-gadget/inspektor-gadget/pkg/datasource/formatters/json"
+	gadgetcontext "github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-context"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+	_ "github.com/inspektor-gadget/inspektor-gadget/pkg/operators/ebpf"
+	ocihandler "github.com/inspektor-gadget/inspektor-gadget/pkg/operators/oci-handler"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators/simple"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/runtime/local"
+)
+
+func do() error {
+	// First of all, we need to a (simple) operator that we'll use to subscribe
+	// to events. Check the operator documentation to learn more about them.
+	// TODO: documentation
+	// Our operator should be the last of the chain to print information after
+	// all operators have been run.
+	const opPriority = 50000
+
+	myOperator := simple.New("myHandler", simple.OnInit(func(gadgetCtx operators.GadgetContext) error {
+		// Subscribe to all datasources and print their output as json the terminal
+		// Check the datasources documentation for more information
+		// TODO: link to documentation
+		for _, d := range gadgetCtx.GetDataSources() {
+			jsonEncoder, _ := igjson.New(d)
+
+			d.Subscribe(func(source datasource.DataSource, data datasource.Data) error {
+				jsonOutput := jsonEncoder.Marshal(data)
+				fmt.Printf("%s\n", jsonOutput)
+				return nil
+			}, opPriority)
+		}
+		return nil
+	}))
+
+	// Then, we create a gadget context instance. This is the glue that connects
+	// all operators together.
+	// Check the documentation for the gadget context here.
+	// TODO: link to documentation
+	gadgetCtx := gadgetcontext.New(
+		context.Background(),
+		"ghcr.io/inspektor-gadget/gadget/trace_open:latest",
+		gadgetcontext.WithDataOperators(ocihandler.OciHandler, myOperator),
+	)
+
+	// After that, we need a runtime, that's the piece that will run our gadget.
+	// Check the runtime documentation here.
+	runtime := local.New()
+	if err := runtime.Init(nil); err != nil {
+		return fmt.Errorf("runtime init: %w", err)
+	}
+
+	// Before running the gadget, let's set some parameters.
+	// TODO: link to documentation with the parameters supported by this gadgert
+	params := map[string]string{
+		// Filter only events from the root user
+		"operator.oci.ebpf.uid": "0",
+	}
```

Comment:
I found it very painful to get it right. What about the `ParamKey` idea I implemented in one of the commits of this PR?

Reply:
I think it's okay to have these - they're similar to how annotations in Kubernetes look like ([example](https://doc.traefik.io/traefik/routing/providers/kubernetes-ingress/#annotations)). But we should of course provide a good documentation around those params and the actual `annotations` we support, though.

The `ParamKey` looks okay, but the downsides are
* you add a hard dependency on the operator (which in these examples is okay, but still)
* you would need to implement them for all we offer to not be more confusing than helpful, so we might as well add that to the API itself
* the key actual depends on how the operator is used; we could later on have operators as layer operators as well as data operators (think wasm, for example - we can have them as part of the gadget image as well as globally installed wasm programs) - in the former case it would be `operator.oci.wasm`, in the latter it would only be `operator.wasm`

By the way, I thought maybe dropping the starting `operator.` might make sense now that we don't have `gadget` params anymore (right now it's only operator + runtime.)

Reply:
I think the issue is not only about the prefix, but in general about the full key. I find it very difficult for a use of the API to understand what are the parameters (and their type) that are available and how this key should be composed. 

I'm wondering if we can offer a more structured way to set parameters, like a Golang structure.

Reply:
How about additionally defining all params as either `ParamPARAMNAME` or `GlobalParamPARAMNAME` with a proper description that includes the type (we mostly have done it already anyway). Combined with the above that would lead to

```
operators.AddParam(params, ocihandler.OciHandler, ocihandler.ParamFoobar, "value")
```

(changed order of paramMap here). Usually the IDE would autocomplete after typing `ocihandler.Param...` and help with what is available.

Reply:
I think we need more discussion on the parameters thing. IMO we can try to implement basic examples so folks can start playing with the API, in the meanwhile we can continue the parameters discussion.

Reply:
Opened #2893 with some thoughts. I haven't have the change to explore this further.